lib/Graph/Graph.cpp

```diff
@@ -1346,37 +1374,40 @@ BatchedReduceAddNode *Function::createBatchedReduceAdd(llvm::StringRef name,
   return addNode(new BatchedReduceAddNode(name, OT, batch, axis));
 }
 
-BatchedReduceAddNode *Function::createBatchedReduceAdd(llvm::StringRef name,
-                                                       NodeValue batch,
-                                                       unsigned_t axis) {
-  auto outDims = getNewShapeWithoutAxis(batch.dims(), axis);
+BatchedReduceAddNode *
+Function::createBatchedReduceAdd(llvm::StringRef name, NodeValue batch,
+                                 llvm::ArrayRef<unsigned_t> axes,
+                                 bool keepDims) {
+
+  ShapeVector outDims(batch.dims().begin(), batch.dims().end());
+  outDims = keepDims ? getNewShapeReducedAxes(outDims, axes)
```

Comment:
I don't think this is the preferred way to handle `keepDims`. This encodes it into the output shape. This means the Interpreter's kernel, LLVM IRGen, etc. would need to inspect the output shape and behave differently based on it.

Instead I think we should create the output shape assuming `keepDims` is false, and then add a reshape to the output if `keepDims` is true. So you would always create the `BRA`/`BRM` with the shape without axes, and then if `keepDims == true` then add in a reshape here using the reduced axes shape.

Does this seem reasonable? The optimizer will get rid of any extra inserts/copies, so there's no performance cost, and it allows there to be a single canonical form of `BRA`/`BRM`.

Reply:
> This means the Interpreter's kernel, LLVM IRGen, etc. would need to inspect the output shape and behave differently based on it.
I thought that has to be done anyway - A node is created with `TypeRef Result` argument which can take any value, as it can be set to just about anything that's compatible,  through that other `createBatchedReduceAdd` version that takes  `TypeRef outTy ` argument.